// @pyroscope/nodejs/dist/cjs/profilers/heap-profiler.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.HeapProfiler = void 0;
const pprof_1 = require("@datadog/pprof");
const debug_1 = __importDefault(require("debug"));
const log = (0, debug_1.default)('pyroscope::profiler::heap');
class HeapProfiler {
    labels;
    lastProfiledAt;
    constructor() {
        this.labels = {};
        this.lastProfiledAt = new Date();
    }
    getLabels() {
        return this.labels;
    }
    profile() {
        log('profile');
        const profile = pprof_1.heap.profile();
        const lastProfileStartedAt = this.lastProfiledAt;
        this.lastProfiledAt = new Date();
        return {
            profile,
            startedAt: lastProfileStartedAt,
            stoppedAt: this.lastProfiledAt,
        };
    }
    setLabels(labels) {
        this.labels = labels;
    }
    start(args) {
        log('start');
        this.lastProfiledAt = new Date();
        pprof_1.heap.start(args.samplingIntervalBytes, args.stackDepth);
    }
    stop() {
        log('stop');
        pprof_1.heap.stop();
        return null;
    }
}
exports.HeapProfiler = HeapProfiler;
