// @pyroscope/nodejs/dist/cjs/profilers/wall-profiler.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WallProfiler = void 0;
const pprof_1 = require("@datadog/pprof");
const debug_1 = __importDefault(require("debug"));
const MICROS_PER_SECOND = 1e6;
const log = (0, debug_1.default)('pyroscope::profiler::wall');
class WallProfiler {
    labels;
    lastProfiledAt;
    lastSamplingIntervalMicros;
    constructor() {
        this.labels = {};
        this.lastProfiledAt = new Date();
    }
    getLabels() {
        return this.labels;
    }
    profile() {
        log('profile');
        return this.innerProfile(true);
    }
    setLabels(labels) {
        this.labels = labels;
    }
    start(args) {
        if (!pprof_1.time.isStarted()) {
            log('start');
            this.lastProfiledAt = new Date();
            this.lastSamplingIntervalMicros = args.samplingDurationMs;
            pprof_1.time.start({
                sourceMapper: args.sourceMapper,
                durationMillis: args.samplingDurationMs,
                intervalMicros: args.samplingIntervalMicros,
                withContexts: true,
                workaroundV8Bug: true,
            });
            pprof_1.time.setContext({});
        }
    }
    stop() {
        log('stop');
        return this.innerProfile(false);
    }
    innerProfile(restart) {
        pprof_1.time.setContext({});
        const profile = pprof_1.time.stop(restart, () => this.labels);
        const lastProfileStartedAt = this.lastProfiledAt;
        this.lastProfiledAt = new Date();
        return {
            profile,
            sampleRate: Math.ceil(MICROS_PER_SECOND / this.lastSamplingIntervalMicros),
            startedAt: lastProfileStartedAt,
            stoppedAt: this.lastProfiledAt,
        };
    }
}
exports.WallProfiler = WallProfiler;
