// @pyroscope/nodejs/dist/esm/profilers/wall-profiler.js
import { time } from '@datadog/pprof';
import debug from 'debug';
const MICROS_PER_SECOND = 1e6;
const log = debug('pyroscope::profiler::wall');
export class WallProfiler {
    labels;
    lastProfiledAt;
    lastSamplingIntervalMicros;
    constructor() {
        this.labels = {};
        this.lastProfiledAt = new Date();
    }
    getLabels() {
        return this.labels;
    }
    profile() {
        log('profile');
        return this.innerProfile(true);
    }
    setLabels(labels) {
        this.labels = labels;
    }
    start(args) {
        if (!time.isStarted()) {
            log('start');
            this.lastProfiledAt = new Date();
            this.lastSamplingIntervalMicros = args.samplingDurationMs;
            time.start({
                sourceMapper: args.sourceMapper,
                durationMillis: args.samplingDurationMs,
                intervalMicros: args.samplingIntervalMicros,
                withContexts: true,
                workaroundV8Bug: true,
            });
            time.setContext({});
        }
    }
    stop() {
        log('stop');
        return this.innerProfile(false);
    }
    innerProfile(restart) {
        time.setContext({});
        const profile = time.stop(restart, () => this.labels);
        const lastProfileStartedAt = this.lastProfiledAt;
        this.lastProfiledAt = new Date();
        return {
            profile,
            sampleRate: Math.ceil(MICROS_PER_SECOND / this.lastSamplingIntervalMicros),
            startedAt: lastProfileStartedAt,
            stoppedAt: this.lastProfiledAt,
        };
    }
}
